Sources/ClientRuntime/EventStream/DefaultMessageEncoderStream.swift

```diff
@@ -0,0 +1,151 @@
+                }
+
+                // marshall event to message
+                let message = try event.marshall(encoder: requestEncoder)
+
+                // sign the message
+                let signedMessage = try await messageSinger.sign(message: message)
+
+                // encode again the signed message
+                let data = try messageEncoder.encode(message: signedMessage)
+                return data
+            }
+        }
+
+        public func makeAsyncIterator() -> AsyncIterator {
+            AsyncIterator(stream: stream,
+                          messageEncoder: messageEncoder,
+                          requestEncoder: requestEncoder,
+                          messageSinger: messageSinger)
+        }
+
+        // MARK: Stream
+
+        public var position: ClientRuntime.Data.Index = 0
+
+        public var length: Int?
+
+        public var isEmpty: Bool = false
+
+        public var isSeekable: Bool = false
+
+        var buffer = Data()
+
+        public func read(upToCount count: Int) throws -> ClientRuntime.Data? {
+            fatalError("read(upToCount:) is not supported by AsyncStream backed streams")
+        }
+
+        public func readToEnd() throws -> ClientRuntime.Data? {
+            fatalError("readToEnd() is not supported by AsyncStream backed streams")
+        }
+
+        public func readAsync(upToCount count: Int) async throws -> Data? {
+            var data = Data()
+            var remaining = count
+
+            // read from buffer
+            if !buffer.isEmpty {
+                let toRead = Swift.min(remaining, buffer.count)
+                data.append(buffer.subdata(in: 0..<toRead))
+
+                // reset buffer to remaining data
+                buffer = Data(buffer.subdata(in: toRead..<buffer.count))
+
+                // update remaining bytes to read
+                remaining -= toRead
+
+                // update position
+                position = position.advanced(by: toRead)
+            }
+
+            while remaining > 0, let next = try await readAsyncIterator?.next() {
+                // read from async stream
+                let toRead = Swift.min(remaining, next.count)
+                data.append(next[0..<toRead])
+                remaining -= toRead
+
+                // update position
+                position += toRead
+
+                // if we have more data, add it to the buffer
+                if next.count > toRead {
+                    buffer.append(next[toRead..<next.count])
+                }
+            }
+
+            // async stream has ended, return nil to mark stream end
+            if data.isEmpty {
+                return nil
+            }
+
+            return data
+        }
+
+        public func write(contentsOf data: ClientRuntime.Data) throws {
+            fatalError("write(contentsOf:) is not supported by AsyncStream backed streams")
+        }
+
+        public func close() throws {
+            // no-op
```

Comment:
Can you provide context as to why this is a no-op? iirc its because there is nothing for us to "close" in CRT.

Reply:
`close` is used when we have read all the data when receiving the response.

In this case, this stream is used for sending the data and we have no use case of that. Closing is marked by calling the finish on the continuation of the async stream which is base stream for it.

Reply:
Makes sense, can you add this as a comment here?